{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c7cadaf-1ced-4222-aa3b-29161b69e26c",
   "metadata": {},
   "source": [
    "# Script para crear archivo .nt y .nq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04a669a2-ac74-48f6-b6e7-8608fe9ece18",
   "metadata": {},
   "source": [
    "## 1. Archivo prostatecancer.nt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "748e3ea2-29ba-4e67-a553-0b656f6aa2da",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "from rdflib import Graph, Literal, Namespace, URIRef\n",
    "from rdflib.namespace import OWL, RDF, RDFS, SKOS, XSD, DC, DCTERMS, VOID\n",
    "\n",
    "# Abrir archivo CSV\n",
    "input_file = csv.DictReader(open(\"prostate_cancer.csv\"))\n",
    "\n",
    "# Crear un objeto Grafo \n",
    "output_graph = Graph()\n",
    "\n",
    "# URI base\n",
    "base_uri = 'https://prostatecancer.es/data/'\n",
    "\n",
    "# Propiedades comunes (FAIR: reusar vocabularios)\n",
    "\n",
    "GO=\"http://www.ebi.ac.uk/QuickGO/term/GO:\"\n",
    "NCIT=\"https://ncit.nci.nih.gov/ncitbrowser/ConceptReport.jsp?dictionary=NCI_Thesaurus&ns=ncit&code=\"\n",
    "REACTOME=\"https://reactome.org/content/detail/\"\n",
    "SIO=\"https://semanticscience.org/resource/SIO_\"\n",
    "SO=\"http://purl.obolibrary.org/obo/SO_\"\n",
    "Ncit=\"http://purl.obolibrary.org/obo/NCIT_\"\n",
    "OGG=\"http://purl.obolibrary.org/obo/OGG_\"\n",
    "RO=\"http://purl.obolibrary.org/obo/RO_\"\n",
    "TXPO=\"http://purl.obolibrary.org/obo/TXPO_\"\n",
    "Go=\"http://purl.obolibrary.org/obo/GO_\"\n",
    "ENSEMBL=\"https://grch37.ensembl.org/Homo_sapiens/Gene/Summary?g=\"\n",
    "UNIPROT=\"https://www.uniprot.org/uniprotkb/\"\n",
    "FMA=\"http://purl.obolibrary.org/obo/FMA_\"\n",
    "MI=\"http://purl.obolibrary.org/obo/MI_\"\n",
    "pubmed=\"http://pubmed.ncbi.nlm.nih.gov/\" \n",
    "\n",
    "# Recorrer cada linea y convertir a triples RDF\n",
    "\n",
    "for row in input_file:\n",
    "    # Gene\n",
    "    gen_name=row['gene']    \n",
    "    gen_uri = OGG + row['id_gene']\n",
    "    # Prostate_cancer -> Predicted_genes\n",
    "    predicted_gene=SIO+\"000987\"    \n",
    "    \n",
    "    if gen_uri != 'NA':\n",
    "        # prostate_cancer->predicted_gene->gen\n",
    "        output_graph.add( (URIRef(base_uri), URIRef(predicted_gene), URIRef(gen_uri)))\n",
    "        \n",
    "        # gen->rdfs:label->nombre\n",
    "        output_graph.add( (URIRef(gen_uri), RDFS.label, Literal(gen_name, lang='en')) )\n",
    "    \n",
    "    #IDs\n",
    "    identifier=SIO+\"000115\"\n",
    "    \n",
    "    ensembl=row['Ensembl']\n",
    "    ensembl_uri=ENSEMBL+ensembl\n",
    "\n",
    "    uniprot=row['Uniprot']\n",
    "    uniprot_uri=UNIPROT+uniprot\n",
    "    \n",
    "    \n",
    "    # gen->identifier->uri(ensembl/Uniprot)\n",
    "    output_graph.add( (URIRef(gen_uri), URIRef(identifier), URIRef(ensembl_uri)) )\n",
    "    output_graph.add( (URIRef(gen_uri), URIRef(identifier), URIRef(uniprot_uri)) )\n",
    "    \n",
    "    # uri(ensembl/Uniprot)->rdfs:label->id (ensembl/Uniprot)\n",
    "    output_graph.add( (URIRef(ensembl_uri), RDFS.label, Literal(ensembl, lang='en')) )\n",
    "    output_graph.add( (URIRef(uniprot_uri), RDFS.label, Literal(uniprot, lang='en')) )\n",
    "    \n",
    "    #Gene description\n",
    "    \n",
    "    description=row['Gene.description']\n",
    "    description_uri=SIO+\"000136\"\n",
    "    output_graph.add( (URIRef(gen_uri), URIRef(description_uri), Literal(description, lang='en')) )\n",
    "    \n",
    "    \n",
    "    #Chromosome\n",
    "    chromosome=row['chromosome']\n",
    "    chromosome_uri=NCIT+row['chromosome_uri']\n",
    "    is_contained_in=SIO+\"000128\"\n",
    "    \n",
    "    output_graph.add( (URIRef(gen_uri), URIRef(is_contained_in), URIRef(chromosome_uri)))\n",
    "    output_graph.add( (URIRef(chromosome_uri), RDFS.label, Literal(chromosome, lang='en')) )\n",
    "    \n",
    "    #Position\n",
    "    position=row['Position']\n",
    "    genomic_position=Ncit+\"C123886\"\n",
    "    output_graph.add( (URIRef(gen_uri), URIRef(genomic_position), Literal(position, lang='en')) )\n",
    "    \n",
    "\n",
    "    # Mutation and type mutation\n",
    "    mutation = row['mutation']\n",
    "    sequence_alteration=SO+\"0001059\"\n",
    "    type_mutation = row['type_mutation']\n",
    "    deletion=SO+\"0000159\"\n",
    "    substitution=SO+\"1000002\"\n",
    "    has=Ncit+\"C101282\"\n",
    "    \n",
    "    if type_mutation == \"Substitution\": \n",
    "        output_graph.add((URIRef(gen_uri),URIRef(substitution),Literal(mutation, lang='en')) )\n",
    "    if type_mutation == \"Deletion\":\n",
    "        output_graph.add((URIRef(gen_uri),URIRef(deletion),Literal(mutation, lang='en')) )\n",
    "        \n",
    "    # Biological process\n",
    "    biological_process=Ncit+\"C17828\"\n",
    "    process=row['biological_process']\n",
    "    id_function=row['id_function']\n",
    "    process_uri=GO+id_function\n",
    "    if id_function != 'NA':\n",
    "        output_graph.add( (URIRef(gen_uri), URIRef(biological_process), URIRef(process_uri)))\n",
    "        output_graph.add( (URIRef(process_uri), RDFS.label, Literal(process, lang='en')) )\n",
    "    \n",
    "    # Cellular component\n",
    "    cellular_component=Go+\"0005575\"\n",
    "    component=row['cellullar.component']\n",
    "    id_component=row['id_component']\n",
    "    component_uri=GO+id_component\n",
    "    if id_component != 'NA':\n",
    "        output_graph.add( (URIRef(gen_uri), URIRef(cellular_component), URIRef(component_uri)))\n",
    "        output_graph.add( (URIRef(component_uri), RDFS.label, Literal(component, lang='en')) )\n",
    "    \n",
    "    # Molecular function\n",
    "    molecular_function=Go+\"0003674\"\n",
    "    molecular=row['molecular_function']\n",
    "    id_molecular=row['id_molecular']\n",
    "    molecular_uri=GO+id_molecular\n",
    "    if id_molecular != 'NA':\n",
    "        output_graph.add( (URIRef(gen_uri), URIRef(molecular_function), URIRef(molecular_uri)))\n",
    "        output_graph.add( (URIRef(molecular_uri), RDFS.label, Literal(molecular, lang='en')) )\n",
    "    \n",
    "    # REACTOME\n",
    "    participes=row['participes_in']\n",
    "    candidate_of=Ncit+\"C138079\"\n",
    "    component_of=TXPO+\"0000039\"\n",
    "    member_of=RO+\"0002350\"\n",
    "    input_of=RO+\"0002352\"\n",
    "    output_of=RO+\"0002353\"\n",
    "    #participes_uri=row['uris_participes_in']\n",
    "    participes_label=row['reactome_label']\n",
    "    reactome_uri=REACTOME+row['reactome_id']\n",
    "    \n",
    "    if participes == \"candidate_of\":\n",
    "        output_graph.add((URIRef(gen_uri), URIRef(candidate_of), URIRef(reactome_uri)))\n",
    "        output_graph.add((URIRef(reactome_uri), RDFS.label, Literal(participes_label,lang='en')))\n",
    "    if participes== \"component_of\":\n",
    "        output_graph.add((URIRef(gen_uri), URIRef(component_of), URIRef(reactome_uri)))\n",
    "        output_graph.add((URIRef(reactome_uri), RDFS.label, Literal(participes_label,lang='en')))   \n",
    "    if participes == \"member_of\":\n",
    "        output_graph.add((URIRef(gen_uri), URIRef(member_of), URIRef(reactome_uri)))\n",
    "        output_graph.add((URIRef(reactome_uri), RDFS.label, Literal(participes_label,lang='en')))\n",
    "    if participes == \"input_of\":\n",
    "        output_graph.add((URIRef(gen_uri), URIRef(input_of), URIRef(reactome_uri)))\n",
    "        output_graph.add((URIRef(reactome_uri), RDFS.label, Literal(participes_label,lang='en')))\n",
    "    if participes == \"output_of\":\n",
    "        output_graph.add((URIRef(gen_uri), URIRef(output_of), URIRef(reactome_uri)))\n",
    "        output_graph.add((URIRef(reactome_uri), RDFS.label, Literal(participes_label,lang='en')))\n",
    "    \n",
    "    # Compartment of reactome\n",
    "    compartment_cell=FMA+\"223408\"\n",
    "    reactome_compartment=row['go_compartment']\n",
    "    go_compartment_uri=row['go_compartment_uri']\n",
    "    reactome_compartment_uri=GO+\"000\"+go_compartment_uri\n",
    "    if go_compartment_uri != 'NA': \n",
    "        output_graph.add((URIRef(reactome_uri), URIRef(compartment_cell), URIRef(reactome_compartment_uri)))\n",
    "        output_graph.add((URIRef(reactome_compartment_uri), RDFS.label, Literal(reactome_compartment,lang='en')))\n",
    "    \n",
    "    # Article\n",
    "    article=NCIT+\"C47902\"\n",
    "    article_uri=pubmed+row['id_article']\n",
    "    article_label=row['name_article']\n",
    "    article_year=row['year_article']\n",
    "    publication_year=MI+\"0886\"\n",
    "    \n",
    "    if gen_name != 'KAT6A':   \n",
    "        output_graph.add((URIRef(gen_uri), URIRef(article), URIRef(article_uri)))\n",
    "        output_graph.add((URIRef(article_uri), RDFS.label, Literal(article_label,lang='en')))\n",
    "        output_graph.add((URIRef(article_uri), URIRef(publication_year), Literal(article_year,lang='en')))\n",
    "\n",
    "output_graph.serialize(destination='prostatecancer.nt', format='nt')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f190d9ec-79da-4934-bbd4-bb567cb5eba8",
   "metadata": {},
   "source": [
    "## 2. Archivo prostatecancer.nq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "5f707cc2-2927-4d99-9375-1d05d578fb12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reemplazos realizados con éxito.\n",
      "Archivo renombrado con éxito.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "nombre_archivo_original = \"prostatecancer.nt\"  # Reemplaza \"archivo.txt\" por la ruta y nombre de tu archivo original\n",
    "nombre_archivo_nuevo = \"prostatecancer.nq\"  # Reemplaza \"nuevo_archivo.txt\" por el nombre que deseas para el archivo modificado\n",
    "\n",
    "# Abrir el archivo original en modo lectura\n",
    "with open(nombre_archivo_original, \"r\") as archivo_original:\n",
    "    lineas = archivo_original.readlines()  # Leer todas las líneas del archivo original\n",
    "\n",
    "# Realizar los reemplazos en cada línea\n",
    "nuevo_contenido = []\n",
    "for linea in lineas:\n",
    "    elementos = linea.rstrip(\"\\n\").split()  # Dividir la línea en elementos separados por espacios\n",
    "    if len(elementos) > 1:  # Verificar que haya más de un elemento en la línea\n",
    "        nuevos_elementos = elementos[:-1]  # Eliminar el último elemento de la línea\n",
    "        nuevos_elementos.append(\"<https://prostatecancer.es/graph/genes_predicted> .\")  # Añadir el nuevo contenido al final de la línea\n",
    "        nueva_linea = \" \".join(nuevos_elementos) + \"\\n\"  # Unir los elementos de la línea con espacios y agregar el carácter de salto de línea\n",
    "        nuevo_contenido.append(nueva_linea)\n",
    "\n",
    "# Abrir el archivo nuevo en modo escritura para guardar los cambios\n",
    "with open(nombre_archivo_nuevo, \"w\") as archivo_nuevo:\n",
    "    archivo_nuevo.writelines(nuevo_contenido)\n",
    "\n",
    "print(\"Reemplazos realizados con éxito.\")\n",
    "\n",
    "# Renombrar el archivo nuevo con el formato deseado\n",
    "nuevo_nombre_con_formato = \"prostatecancer.nq\" # Reemplaza \"nuevo_archivo.txt\" por el nuevo nombre y formato deseado, por ejemplo: \"nuevo_archivo.csv\"\n",
    "os.rename(nombre_archivo_nuevo, nuevo_nombre_con_formato)\n",
    "\n",
    "print(\"Archivo renombrado con éxito.\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3.6",
   "language": "python",
   "name": "python3.6"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
